### PR TITLE
#300 mpy

### PR DESCRIPTION
**Pull request by dsh33782**
I think I've cleaned up conflicts and have updates ready to be merged in dev main.  I ran build ok on my local pc.

### CODE REVIEW

File: src/pz390.java [hunk 1]
```diff
@@ -8129,12 +8133,40 @@ private void ins_B9XX(){
...
+			 big_int1 = new BigInteger(get_signed_bytes(reg_byte, rf2, 8)); 
+																			
+			big_int2 = new BigInteger(get_signed_bytes(reg_byte, rf3, 8));
+			big_int1 = big_int1
+			         .multiply(big_int2);
+
+       fp_bi_to_wreg(reg_byte,rf1,big_int1, 8);
+			 if (big_int1.compareTo(bi_max_pos_long) == 1
+			     || big_int1.compareTo(bi_min_neg_long) == -1) {                  
+			     psw_cc = psw_cc3;
```
Comment (jyganci) [inline src/pz390.java:8161]: Add ``set_psw_check(psw_pic_fx_ovf);`` after setting psw_cc to psw_cc3.

File: src/pz390.java [hunk 2]
```diff
@@ -8129,12 +8133,40 @@ private void ins_B9XX(){
...
+			 if (big_int1.compareTo(bi_max_pos_long) == 1
+			     || big_int1.compareTo(bi_min_neg_long) == -1) {                  
+			     psw_cc = psw_cc3;                          
+			 } else if (big_int1.compareTo(BigInteger.ZERO) == 1){
+                                  psw_cc = psw_cc2;
+                              } else if (big_int1.compareTo(BigInteger.ZERO) == -1){
+                                  psw_cc = psw_cc1;
+                              } else {
+                                  psw_cc = psw_cc0;
+			}
```
Comment (jyganci) [inline src/pz390.java:8168]: Suggestion only. You could replace this else if block with
```
			} else {
				switch (big_int1.compareTo(BigInteger.ZERO))
				{
				case 1:
					psw_cc = psw_cc2;
					break;
				case -1:
					psw_cc = psw_cc1;
					break;
				default:
					psw_cc = psw_cc0;
					break;
				}			
```
Removes one of the big_int1.compareTo() invocations.

File: src/pz390.java [hunk 3]
```diff
@@ -8233,7 +8265,20 @@ private void ins_B9XX(){
 				psw_cc = get_int_log_sub_cc();
 				break;		
 	       case 0xFD: // "B9FD=MSRKC,54,340", // B9FD rrfa MSRKC R1,R2,R3 RPI 2202
+		          psw_check = false;
 	    	      ins_setup_rrfa();
+				  rlv1 = (long)reg.getInt(rf2 + 4) * (long)reg.getInt(rf3 + 4);
+				  rv1  = (int)rlv1;
+				  reg.putInt(rf1 + 4, rv1);
+				  if (rlv1 != (long)rv1){
+					  psw_cc = psw_cc3;
```
Comment (jyganci) [inline src/pz390.java:8274]: Add ``set_psw_check(psw_pic_fx_ovf);`` after setting psw_cc to psw_cc3.

File: src/pz390.java [hunk 4]
```diff
@@ -9233,6 +9278,22 @@ private void ins_E3XX(){   // RPI 1207
...
 			reg.putInt(rf1 + 4, reg.getInt(rf1 + 4) * mem.getInt(xbd2_loc));
 			break;
+		case 0x53: // "E353" "MSC" "RXY" // #300
+		    psw_check = false;
+		    ins_setup_rxy();
+			rlv1 = (long)reg.getInt(rf1 + 4) * (long)mem.getInt(xbd2_loc);
+			rv1  = (int)rlv1;
+			reg.putInt(rf1 + 4, rv1);
+			if (rlv1 != (long)rv1){
+			  psw_cc = psw_cc3;
```
Comment (jyganci) [inline src/pz390.java:9288]: Add ``set_psw_check(psw_pic_fx_ovf);`` after setting psw_cc to psw_cc3.

File: src/pz390.java [hunk 5]
```diff
@@ -9436,23 +9497,32 @@ private void ins_E3XX(){   // RPI 1207
...
+			big_int1 = new BigInteger(get_signed_bytes(reg_byte, rf1, 8)); 
+																			
+			big_int2 = new BigInteger(get_signed_bytes(mem_byte, xbd2_loc, 8));
+			big_int1 = big_int1
+			         .multiply(big_int2);
+			rlv1 = big_int1.longValue();
+			fp_bi_to_wreg(reg_byte,rf1,big_int1, 8);
+			if (big_int1.compareTo(bi_max_pos_long) == 1
+			|| big_int1.compareTo(bi_min_neg_long) == -1) {                  
+                psw_cc = psw_cc3;
```
Comment (jyganci) [inline src/pz390.java:9509]: Add ``set_psw_check(psw_pic_fx_ovf);`` after setting psw_cc to psw_cc3.

File: src/pz390.java [hunk 6]
```diff
@@ -13875,16 +13945,24 @@ private int get_long_log_sub_cc() {
...
-		byte[] new_byte = new byte[data_len + 1];
-		System.arraycopy(data_byte, data_offset, new_byte, 1, data_len); // RPI
-																			// 411
+		byte[] new_byte = new byte[data_len + 1]; // default all 0x00
+		System.arraycopy(data_byte, data_offset, new_byte, 1, data_len); // RPI 411
+		return new_byte;
+	}
+	private byte[] get_signed_bytes(byte[] data_byte, int data_offset, int data_len) {  // #300
+
+		 * 2022-02-21 dsh issue 300 fix to support 2s compliment negative numbers
```
Comment (jyganci) [inline src/pz390.java:13961]: "compliment" should be "complement". There may be other occurrences.

File: tests/TESTINS5.MLC [hunk 7]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+* MS
+         L      R1,=F'2'
+         MS    R1,=F'3'
+         CL     R1,=F'6'
+         RT1  CC0
+         L       R1,=F'2'
+         MS    R1,=F'-3'
+         CL     R1,=F'-6'
+         RT1  CC0
+         L       R1,=F'-2'
```
Comment (jyganci) [inline tests/TESTINS5.MLC:412]: Add the (-2) * 3 test.

File: tests/TESTINS5.MLC [hunk 8]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         RT1    CC0 
+* MS
+         L      R1,=F'2'
+         MS    R1,=F'3'
+         CL     R1,=F'6'
+         RT1  CC0
+         L       R1,=F'2'
+         MS    R1,=F'-3'
+         CL     R1,=F'-6'
+         RT1  CC0
```
Comment (jyganci) [inline tests/TESTINS5.MLC:411]: Add the (-2) * 3 test.

File: tests/TESTINS5.MLC [hunk 9]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+* MSY
+         L      R1,=F'2'
+         MSY    R1,=F'3'
+         CL     R1,=F'6'
+         RT1  CC0
+         L       R1,=F'2'
+         MSY    R1,=F'-3'
+         CL     R1,=F'-6'
+         RT1  CC0
```
Comment (jyganci) [inline tests/TESTINS5.MLC:424]: Add the (-2) * 3 test.

File: tests/TESTINS5.MLC [hunk 10]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+          RT1  CC0
+* MSGRKC
+         LMG   R5,R6,=FL8'2,3'
+         SR    R0,R0
+         MSGRKC  R2,R5,R6
+         RT1    CC2
+         CLG    R2,=FL8'6'
+         RT1    CC0 
+         LMG   R5,R6,=FL8'2,-3'
+         MSGRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:437]: Add ``RT1   CC1`` after the MSGRKC.

File: tests/TESTINS5.MLC [hunk 11]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CLG    R2,=FL8'-6'
+         RT1    CC0 
+         LMG   R5,R6,=FL8'-2,3'
+         MSGRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:441]: Add ``RT1   CC1`` after the MSGRKC.

File: tests/TESTINS5.MLC [hunk 12]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CLG    R2,=FL8'-6'
+         RT1    CC0 
+         LMG   R5,R6,=FL8'-2,-3'
+         MSGRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:445]: Add ``RT1   CC2`` after the MSGRKC.

File: tests/TESTINS5.MLC [hunk 13]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CLG   R2,=FL8'6'
+         RT1    CC0 
+         LMG   R5,R6,=XL8'7FFFFFFFFFFFFFFF,2'
+         MSGRKC  R2,R5,R6
+          RT1    CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:450]: Even though overflow occurs, GR2 is still set to a value. You might want to add
```
         CLG   R2,=FL8'-2'
         RT1   CC0
```
You should, of course, compute the expected value yourself. I did it in Eclipse. Code snippet:
```
        bi = new BigInteger("9223372036854775807");
        ba = bi.toByteArray();
        System.out.println("bi "+bi.toString());
        printByteArray("ba", ba);
        
        bi = bi.multiply(new BigInteger("2"));
        ba = bi.toByteArray();
        System.out.println("bi * 2 = "+bi.toString());
        printByteArray("ba", ba);
        
        ba[0] = (byte)0xff;
        bi = new BigInteger(ba);
        ba = bi.toByteArray();
        System.out.println("bi re-build = "+bi.toString());
        printByteArray("ba", ba);
```
and resulting output
```
bi 9223372036854775807
ba: 7f ff ff ff ff ff ff ff 
bi * 2 = 18446744073709551614
ba: 00 ff ff ff ff ff ff ff fe 
bi re-build = -2
ba: fe 
```
Here's the printByteArray() submodule:
```
    static void printByteArray(String name, byte[] b)
    {
        System.out.print(name+": ");
        for (int i = 0; i < b.length; i++)
            System.out.printf("%02x ",b[i]);
        System.out.println();
    }
```

File: tests/TESTINS5.MLC [hunk 14]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         LMG   R5,R6,=FL8'-2,-3'
+         MSGRKC  R2,R5,R6
+         CLG   R2,=FL8'6'
+         RT1    CC0 
+         LMG   R5,R6,=XL8'7FFFFFFFFFFFFFFF,2'
+         MSGRKC  R2,R5,R6
+          RT1    CC3  
+* MSRKC
+         LM    R5,R6,=F'2,3'
+         MSRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:453]: Add ``RT1   CC2`` after the MSRKC.

File: tests/TESTINS5.MLC [hunk 15]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         RT1    CC0 
+         LMG   R5,R6,=FL8'-2,-3'
+         MSGRKC  R2,R5,R6
+         CLG   R2,=FL8'6'
+         RT1    CC0 
+         LMG   R5,R6,=XL8'7FFFFFFFFFFFFFFF,2'
+         MSGRKC  R2,R5,R6
+          RT1    CC3  
+* MSRKC
+         LM    R5,R6,=F'2,3'
```
Comment (jyganci) [inline tests/TESTINS5.MLC:452]: Preset CC before the MSRKC to something other than 2.

File: tests/TESTINS5.MLC (hunk 13 above)
Comment (jyganci) [reply tests/TESTINS5.MLC:450]: For completeness, you might add MSGRKC tests for 2 * 0, 0 * 2, (-2) * 0, 0 * (-2), and 0 * 0.

File: tests/TESTINS5.MLC [hunk 16]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         MSRKC  R2,R5,R6
+         CL      R2,=F'6'
+         RT1    CC0 
+         LM     R5,R6,=F'2,-3'
+         MSRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:457]: Add ``RT1   CC1`` after the MSRKC.

File: tests/TESTINS5.MLC [hunk 17]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CL      R2,=F'-6'
+         RT1    CC0 
+         LM     R5,R6,=F'-2,3'
+         MSRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:461]: Add ``RT1   CC1`` after the MSGRKC.

File: tests/TESTINS5.MLC [hunk 18]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CL      R2,=F'-6'
+         RT1    CC0 
+         LM     R5,R6,=F'-2,-3'
+         MSRKC  R2,R5,R6
```
Comment (jyganci) [inline tests/TESTINS5.MLC:465]: Add ``RT1   CC2`` after the MSRKC.

File: tests/TESTINS5.MLC [hunk 19]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         CL     R2,=F'6'
+         RT1    CC0 
+         LM   R5,R6,=XL4'7FFFFFFF,2'
+         MSRKC  R2,R5,R6
+         RT1    CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:470]: Same comments as for MSGRKC. Even thouh overflow occurs, check that R2 is set to the expected result.

Also, add the tests for a product of zero.

File: tests/TESTINS5.MLC [hunk 20]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         LM     R5,R6,=F'-2,-3'
+         SR      R0,R0
+         MSRKC  R2,R5,R6
+         RT1   CC2
+         CL     R2,=F'6'
+         RT1    CC0 
+         LM   R5,R6,=XL4'7FFFFFFF,2'
+         SR       R0,R0
+         MSRKC  R2,R5,R6
+         RT1    CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:529]: Repeated MSGRKC, MSRKC tests???

File: tests/TESTINS5.MLC [hunk 21]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+	RT1	CC0
+         L	R2,=F'-2'
+         SR       R0,R0
+	MSC R2,=F'-3'
+          RT1  CC2
+          CL    R2,=F'6'
+	RT1	CC0
+         L	R2,=X'7FFFFFFF'
+	MSC R2,=F'2'
+          RT1  CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:557]: Same comments as for MSGRKC. Even though overflows, you should check that R2 is set to the expected value.

Also add tests for product equal to zero.

File: tests/TESTINS5.MLC [hunk 22]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+* MSGC
+         LG	R2,=FL8'2'
+         SR       R0,R0
+	MSC R2,=F'2'
+          RT1  CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:562]: Delete these lines???

File: tests/TESTINS5.MLC [hunk 23]
```diff
@@ -341,6 +374,221 @@ BIC_OK2  EQU    *
...
+         LG	R2,=FL8'-2'
+         SR       R0,R0
+	MSGC R2,=FL8'-3'
+          RT1  CC2
+          CLG    R2,=FL8'6'
+	RT1	CC0
+         LG	R2,=X'7FFFFFFFFFFFFFFF'
+         SR       R0,R0
+	MSGC R2,=FL8'2'
+          RT1  CC3
```
Comment (jyganci) [inline tests/TESTINS5.MLC:591]: Same comments about overflow. Should still check that R2 contains the expected value.

Also add tests for product equal to zero.

File: src/pz390.java (hunk 2 above)
Comment (jyganci) [reply src/pz390.java:8168]: I had a typo in the switch statement -- big_int2 instead of the correct big_int1. Fixed the comment, but added this one in case you already looked at it. Sorry about that.
